packages/edit-post/src/components/header/fullscreen-mode-close/index.js

```diff
@@ -7,37 +7,64 @@ import { get } from 'lodash';
  * WordPress dependencies
  */
 import { useSelect } from '@wordpress/data';
-import { Button } from '@wordpress/components';
+import { Button, Icon } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 import { addQueryArgs } from '@wordpress/url';
 import { wordpress } from '@wordpress/icons';
 
 function FullscreenModeClose() {
-	const { isActive, postType } = useSelect( ( select ) => {
-		const { getCurrentPostType } = select( 'core/editor' );
-		const { isFeatureActive } = select( 'core/edit-post' );
-		const { getPostType } = select( 'core' );
+	const { isActive, isRequestingSiteIcon, postType, siteIconUrl } = useSelect(
+		( select ) => {
+			const { getCurrentPostType } = select( 'core/editor' );
+			const { isFeatureActive } = select( 'core/edit-post' );
+			const { isResolving } = select( 'core/data' );
+			const { getEntityRecord, getPostType } = select( 'core' );
+			const siteData =
+				getEntityRecord( 'root', '__unstableBase', undefined ) || {};
 
-		return {
-			isActive: isFeatureActive( 'fullscreenMode' ),
-			postType: getPostType( getCurrentPostType() ),
-		};
-	}, [] );
+			return {
+				isActive: isFeatureActive( 'fullscreenMode' ),
+				isRequestingSiteIcon: isResolving( 'core', 'getEntityRecord', [
+					'root',
+					'__unstableBase',
+					undefined,
+				] ),
+				postType: getPostType( getCurrentPostType() ),
+				siteIconUrl: siteData.site_icon_url,
+			};
+		},
+		[]
+	);
 
 	if ( ! isActive || ! postType ) {
 		return null;
 	}
 
+	let buttonIcon = <Icon size="36px" icon={ wordpress } />;
+
+	if ( siteIconUrl ) {
+		buttonIcon = (
+			<img
+				alt={ __( 'Site Icon' ) }
+				className="edit-post-fullscreen-mode-close_site-icon"
+				src={ siteIconUrl }
+			/>
+		);
+	} else if ( isRequestingSiteIcon ) {
+		buttonIcon = null;
+	}
+
 	return (
 		<Button
-			className="edit-post-fullscreen-mode-close"
-			icon={ wordpress }
-			iconSize={ 36 }
+			className="edit-post-fullscreen-mode-close has-icon"
```

Comment:
I might be missing some context, but is there a reason why we're adding the `.has-icon` class?

Reply:
Good question. 

Previously, we used to pass the `wordpress` icon (an [`SVG` component](https://github.com/WordPress/gutenberg/blob/41cf77e0fbc52d4b57c1a477d8754dafbb0f4ddf/packages/icons/src/library/wordpress.js#L7-L9) from [`@wordpress/icons`](https://github.com/WordPress/gutenberg/pull/22952/files/c0b49d2a273eeeb4b2c7d731ce150bb171ef9c66#diff-5e54b8ed48d99cb46a591a186b9ad138L13)) via the [`icon` prop](https://github.com/WordPress/gutenberg/pull/22952/files/c0b49d2a273eeeb4b2c7d731ce150bb171ef9c66#diff-5e54b8ed48d99cb46a591a186b9ad138L34) to `Button`. The `Button` component forwards that prop to an [`Icon` element it embeds](https://github.com/WordPress/gutenberg/blob/41cf77e0fbc52d4b57c1a477d8754dafbb0f4ddf/packages/components/src/button/index.js#L113), and sets the [`has-icon` prop](https://github.com/WordPress/gutenberg/blob/41cf77e0fbc52d4b57c1a477d8754dafbb0f4ddf/packages/components/src/button/index.js#L63).

Since we're no longer passing the logo to `Button` via the `icon` prop (but wrap the Button around the logo), we're losing the `.has-icon` styling and have to add it back manually.

However, we might explore an alternative...

Reply:
The `Icon` component [seems to accept](https://github.com/WordPress/gutenberg/blob/bdfdd73f4bd9295057b9bfa51fc996185dedb44f/packages/components/src/icon/index.js#L62-L67) pretty much any valid component as a fallback. So we might be able to continue to pass in our custom log (or fallback) via `Button`'s `icon` prop, and remove the manually added `has-icon` prop (which is arguably an implementation thing rather than part of the API).

Reply:
I was experimenting a bit with this, and it would be possible. However, the `Icon` component is also passing a `size` prop to the element it clones, and `<img />` elements don't support that. So let's keep things as they are.